### PR TITLE
migration: enable split migration job creation

### PR DESCRIPTION
**Pull request by bootjp**
## Summary
- keep StartSplitMigration fail-closed until an explicit migration capability gate succeeds
- create planned SplitJob records through the catalog coordinator when the gate is open
- reject reserved migration/control ranges and enforce the one-live-job limit before writing migration state

## Tests
- go test ./adapter -run 'TestDistributionServer(StartSplitMigration|SplitJobRPCs)' -count=1
- go test ./distribution -count=1
- golangci-lint --config=.golangci.yaml run ./adapter ./distribution --timeout=5m

Note: a broader go test ./adapter ./distribution -count=1 run previously timed out in adapter after 10m during wider raft/gRPC tests; the targeted DistributionServer tests and distribution package pass.

### CODE REVIEW

**Comment by bootjp** [comment]
@codex review

**Comment by bootjp** [comment]
Addressed latest split migration readiness feedback in 53d38e31d0e8f9fa32a76a970218c0c32f7c44af.

Changes:
- make pinReadTS return a releasable empty token when the tracker or server is nil
- simplify split migration capability gate errors so the gate result is returned directly
- add regression coverage for tracker-less read pins and gate error propagation

Validation:
- go test ./adapter -run 'TestDistributionServer(StartSplitMigration|PinReadTS|SplitJobRPCs)' -count=1\n- go test ./distribution -count=1\n- golangci-lint --config=.golangci.yaml run ./adapter ./distribution --timeout=5m\n- git diff --check\n\n@codex review

**Comment by bootjp** [comment]
Addressed the current split migration start-guard feedback in fa769ac5ccb1c888cd55d16134af78ce9cbc5a13.

Changes:
- include catalog version and source route records in the split-job creation OCC read set
- reject StartSplitMigration requests whose target group is not one of this node's configured raft groups
- gate Distribution/StartSplitMigration during startup rotation with the same mutator gate as SplitRange

Validation:
- go test ./adapter -run 'TestDistributionServerStartSplitMigration|TestDistributionServerSplitJobRPCs' -count=1\n- go test . -run 'TestStartupPublicKVGate|TestStartupGatedCoordinator' -count=1\n- go test ./distribution -count=1\n- golangci-lint --config=.golangci.yaml run ./adapter . --timeout=5m\n- git diff --check\n\n@codex review

**Comment by bootjp** [comment]
Addressed the remaining split migration source-state guard in 4447dd315213282d49714e9945f52aa2c0c08f3b.

Changes:
- reject StartSplitMigration when the source route is not RouteStateActive
- add coverage for write-fenced, migrating-source, and migrating-target source routes

Validation:
- go test ./adapter -run 'TestDistributionServerStartSplitMigration|TestDistributionServerPinReadTS' -count=1\n- go test ./distribution -run 'TestMigration|TestInitializeSplitJobPlan|TestSplitJob' -count=1\n- golangci-lint --config=.golangci.yaml run ./adapter ./distribution --timeout=5m\n- git diff --check\n\n@codex review

**Comment by bootjp** [comment]
Addressed the current CI failure in f22c4ff9edd3d778a18d285432f7261f42394ba4.\n\nChanges:\n- make the urgent delta-compactor pagination test call the targeted urgent compaction loop synchronously\n- remove the flaky Run-loop observation where the local test coordinator could expose the base-meta update before all delete elems were applied under the race detector\n\nValidation:\n- go test -race ./adapter -run '^TestDeltaCompactor_UrgentCompactionPagination$' -count=30\n- go test ./adapter -run '^TestDeltaCompactor_' -count=1\n- go test -race ./adapter -run '^(TestDeltaCompactor_|TestDistributionServerStartSplitMigration|TestDistributionServerPinReadTS|TestDistributionServerSplitJobRPCs)' -count=1\n- go test ./distribution -run 'TestMigration|TestInitializeSplitJobPlan|TestSplitJob' -count=1\n- git diff --check\n\n@codex review

**Comment by bootjp** [comment]
Current-head review refresh requested.

Latest head f22c4ff9edd3d778a18d285432f7261f42394ba4 already includes the prior start-migration guard fixes:
- pinReadTS returns a non-nil releasable token when the tracker is absent, covered by TestDistributionServerPinReadTSWithoutTrackerReturnsReleasableToken
- non-active source routes and unknown target groups are rejected before job creation
- StartSplitMigration is included in the startup rotation mutating-RPC gate

CI is green on the latest head.

@codex review

**Comment by bootjp** [comment]
Addressed the current-head split migration capability-gate wiring feedback in ce0c2e1b4ed48e7c3f8a2708a580bd1dd6dea1ed.

Changes:
- wire WithSplitMigrationCapabilityGate into the production DistributionServer construction
- keep the existing target-group, catalog, source-state, and startup-rotation guards in place
- add a main-package regression test for the ready split migration capability gate

Validation:
- go test ./adapter -run 'TestDistributionServerStartSplitMigration|TestDistributionServerPinReadTS|TestDistributionServerSplitJobRPCs' -count=1 -timeout=180s\n- go test . -run 'TestSplitMigrationCapabilityGateIsReady|TestStartupPublicKVGate_BlocksMutatorsUntilReady' -count=1 -timeout=180s\n- go test ./distribution -run 'TestMigration|TestInitializeSplitJobPlan|TestSplitJob' -count=1 -timeout=180s\n- go test -run '^$' ./... -count=1 -timeout=300s\n- golangci-lint --config=.golangci.yaml run . ./adapter ./distribution --timeout=5m\n- git diff --check\n\nCaller audit:\n- rg -n "StartSplitMigration\\(|verifySplitMigrationCapability\\(|WithSplitMigrationCapabilityGate\\(|splitMigrationCapabilityGate|WithDistributionKnownRaftGroups\\(" . -g'*.go'\n\n@codex review

**Comment by bootjp** [comment]
Addressed the current-head split migration capability-gate feedback in 49adab3dd41acf0042af2724ac2f5f8dc47291f6.

Changes:
- add Distribution/GetSplitMigrationCapability as a read-only migration capability probe
- make the production StartSplitMigration gate fan out to the configured voter peers and fail closed if any peer is missing, unreachable, or not capable
- keep nil/no-peer configurations fail-closed and cover peer success, failure, and no-peer cases

Validation:
- make -C proto gen
- go test ./adapter -run 'TestDistributionServer(GetSplitMigrationCapability|StartSplitMigration|PinReadTS|SplitJobRPCs)' -count=1 -timeout=180s\n- go test . -run 'TestSplitMigrationCapability|TestStartupPublicKVGate_BlocksMutatorsUntilReady' -count=1 -timeout=180s\n- go test ./distribution -run 'TestMigration|TestInitializeSplitJobPlan|TestSplitJob' -count=1 -timeout=180s\n- go test -run '^$' ./... -count=1 -timeout=300s\n- golangci-lint --config=.golangci.yaml run . ./adapter ./distribution --timeout=5m\n- git diff --check\n\nCaller audit:\n- rg -n "StartSplitMigration\\(|verifySplitMigrationCapability\\(|WithSplitMigrationCapabilityGate\\(|newSplitMigrationCapabilityGate\\(|probeSplitMigrationCapabilityPeer\\(|GetSplitMigrationCapability\\(|splitMigrationCapabilityPeers\\(" . -g'*.go'\n\n@codex review

**Comment by bootjp** [comment]
Addressed the current-head split migration capability-gate follow-up in c64ff5fef248e4c6d94f69b3c1a00965dfac7001.

Changes:
- keep Distribution/GetSplitMigrationCapability fail-closed until a split-job runner is registered
- move the external capability fanout before the catalog mutex is acquired
- let the production gate resolve the current voter set at probe time instead of pinning the startup bootstrap seed

Validation:
- go test ./adapter -run 'TestDistributionServer(GetSplitMigrationCapability|StartSplitMigration|PinReadTS|SplitJobRPCs)' -count=1 -timeout=180s\n- go test . -run 'TestSplitMigrationCapability|TestStartupPublicKVGate_BlocksMutatorsUntilReady' -count=1 -timeout=180s\n- golangci-lint --config=.golangci.yaml run . ./adapter --timeout=5m\n- git diff --check HEAD~1..HEAD\n\nCaller audit:\n- git show --name-only HEAD confirms the change is limited to DistributionServer capability handling and main-package gate wiring/tests.\n\n@codex review

**Comment by bootjp** [comment]
Fixed the current ubuntu test failure in a775fdaaa41f8073a77a725e17ddaababa01a320.

Cause:
- `go test -race ./...` timed out in adapter while `TestLuaNegativeTypeCache_BoundedSize` was running for more than 9 minutes. The test was doing 10k+ real missing-key storage probes even though the invariant under test is the in-memory negative-cache bound.

Changes:
- kept `keyType` behavior unchanged by extracting the negative-cache insertion into `rememberNegativeType`.
- changed the bounded-size regression to exercise that in-memory bound directly, while leaving the wire-level Lua tests and single-probe test to cover real storage probing.

Validation:
- go test ./adapter -run 'TestLuaNegativeTypeCache_BoundedSize|TestLuaNegativeTypeCache_SingleProbePerKey' -count=1 -timeout=180s\n- go test -race ./adapter -run 'TestLuaNegativeTypeCache_BoundedSize|TestLuaNegativeTypeCache_SingleProbePerKey' -count=1 -timeout=180s\n- go test ./adapter -run 'TestDistributionServer(GetSplitMigrationCapability|StartSplitMigration|PinReadTS|SplitJobRPCs)' -count=1 -timeout=180s\n- go test . -run 'TestSplitMigrationCapability|TestStartupPublicKVGate_BlocksMutatorsUntilReady' -count=1 -timeout=180s\n- golangci-lint --config=.golangci.yaml run ./adapter . --timeout=5m\n- git diff --check\n\n@codex review

**Comment by bootjp** [comment]
Addressed the current-head split migration capability feedback in c29806025400c0896df536c8ab1eca6b2038284c.

Changes:
- advertise the documented cap_migration_v2 token
- mark the production DistributionServer as split-job-runner ready now that this PR enables StartSplitMigration job creation
- build the readiness peer set from every configured raft group runtime and include learners, with cross-group peer dedupe

Validation:
- go test ./adapter -run 'TestDistributionServer(GetSplitMigrationCapability|StartSplitMigration|PinReadTS|SplitJobRPCs)' -count=1 -timeout=180s
- go test . -run 'TestSplitMigrationCapability|TestStartupPublicKVGate_BlocksMutatorsUntilReady' -count=1 -timeout=180s
- go test . -run 'TestSplitMigrationCapability' -count=1 -timeout=180s
- golangci-lint --config=.golangci.yaml run . ./adapter --timeout=5m
- git diff --check

Caller audit:
- checked StartSplitMigration, GetSplitMigrationCapability, WithSplitMigrationCapabilityGate, WithSplitJobRunnerReady, and splitMigrationCapabilityPeerSource call sites; production wiring is limited to main.go and tests cover adapter/main behavior.

@codex review

**Comment by bootjp** [comment]
Addressed the current-head split migration capability findings in 70fb630f55ee3e4d3a77ae855c4ab21ec8efa3fc.

Changes:
- kept production split migration capability closed until a real split-job runner is wired
- made capability probing require the documented cap_migration_v2 token, not just migration_capable=true
- fail closed when any configured Raft group reports zero migration capability peers
- made read-pin release explicit and nil-safe for StartSplitMigration/SplitRange

Validation:
- go test . -run 'TestSplitMigrationCapability|TestProbeSplitMigrationCapability|TestDistributionCatalogGroupID|TestSetupDistributionCatalog' -count=1 -timeout=180s
- go test ./adapter -run 'TestDistributionServer(GetSplitMigrationCapability|StartSplitMigration|SplitRange)' -count=1 -timeout=180s
- go test . -count=1 -timeout=180s
- golangci-lint --config=.golangci.yaml run . ./adapter --timeout=5m
- git diff --check

Caller audit:
- rg -n "StartSplitMigration|GetSplitMigrationCapability|WithSplitMigrationCapabilityGate|WithSplitJobRunnerReady|splitJobRunnerReady|verifySplitMigrationCapability|verifyStartSplitMigration|splitMigrationCapabilityPeerSourceForRuntimes|splitMigrationCapabilityPeersFromConfiguration|probeSplitMigrationCapabilityPeer|SplitMigrationCapabilityV2" adapter main.go main_catalog_test.go -g"*.go"

Note: go test . ./adapter -count=1 -timeout=240s completed the root package, then timed out in the existing adapter package-wide suite; the changed adapter paths passed the focused tests above.

@codex review

File: main.go [hunk 1]
```diff
@@ -498,11 +500,19 @@ func run() error {
...
+		splitMigrationCapabilityProbeTimeout,
+		nil,
+	)
 	distServer := adapter.NewDistributionServer(
 		cfg.engine,
 		distCatalog,
 		adapter.WithDistributionCoordinator(coordinate),
 		adapter.WithDistributionActiveTimestampTracker(readTracker),
+		adapter.WithDistributionKnownRaftGroups(shardGroupIDs(shardGroups)...),
+		adapter.WithSplitMigrationCapabilityGate(splitMigrationGate),
```
[the remark this answers is not in the document]
Comment (bootjp) [reply main.go:514]: Superseded by the later fail-closed review: production now wires the cluster capability gate but intentionally does not pass WithSplitJobRunnerReady until a real runner is started, so peers keep migration_capable=false.
